### PR TITLE
fix(portal): popper is updated when container height is updated

### PR DESCRIPTION
**Pull request by maltja**
The popper needs to recalculate on which side the container should flip at when the container's
dimension changes

fix Lundalogik/crm-feature#1722
fix Lundalogik/crm-feature#1705

Review:
- [x] Commits are [atomic](https://seesparkbox.com/foundry/atomic_commits_with_git)
- [x] Commits have the correct *type* for the changes made
- [x] Commits with *breaking changes* are marked as such

### Browsers tested:
(Check any that applies, it's ok to leave boxes unchecked if testing something didn't seem relevant.)

Windows:
- [ ] Chrome
- [x] Edge
- [x] Firefox

Linux:
- [ ] Chrome
- [ ] Firefox

macOS:
- [x] Chrome
- [ ] Firefox
- [ ] Safari

Mobile:
- [x] Chrome on Android
- [x] iOS

### CODE REVIEW

File: src/components/picker/picker.tsx [hunk 1]
```diff
@@ -98,6 +97,9 @@ export class Picker {
 
     /**
      * True if the dropdown list should be displayed without cutting the content
+     *
+     * @deprecated This functiallity dosen't work anymore and should be
+     * handled by the searcher insted
```
Comment (adrianschmidt) [inline src/components/picker/picker.tsx:102]: Small update on this. The picker wasn't actually limiting the number of items in the list, and so this property wasn't to avoid that. Instead, this property was there to allow the dropdown to become high enough to show all items in the list, regardless of whether that made the dropdown extend outside the screen. Thus, if not all items fit on screen, the user would scroll the whole page instead of the content of the dropdown.

It seems this hasn't worked lately though. Compare the _With icons and displaying full list without cutting content_ example of version [30.0.1](http://lime-documentation-lime-elements.s3-website-eu-west-1.amazonaws.com/versions/30.0.1/#/component/limel-picker/) (working) and version [30.0.2](http://lime-documentation-lime-elements.s3-website-eu-west-1.amazonaws.com/versions/30.0.2/#/component/limel-picker/) (broken). Type an `a` in the input to get a long list of results.

As we found previously, this property was originally added in commit 7f0a31a9e51f76acb866058d7aadde43771e0cd7, but I finally managed to find the PR for that commit: #393, fixing issue #363.

We should test the stuff listed in that issue, to make sure we don't reintroduce any of those problems.

Comment (adrianschmidt) [reply src/components/picker/picker.tsx:102]: > to make sure we don't reintroduce any of those problems

(I don't think we will, to be fair. There's a lot of things that are different about the dropdown now: we calculate the size, we can pop it either above or below the trigger, you can scroll the page while the dropdown is visible, etc.)

Comment (adrianschmidt) [reply src/components/picker/picker.tsx:102]: Oh, yeah, what I was _going_ to say was that we should update this comment a bit. Perhaps something like

```suggestion
     * @deprecated This was used for a workaround, and isn't needed any
     * longer. Setting it has no effect, and the property will be removed
     * in the next major version.
```

Comment (maltja) [reply src/components/picker/picker.tsx:102]: nice that you found the issues! 🎉  For me both of the versions are broking, I don't know if I do anything wrong when I tested them. 

"If not all items fit on screen, the user would scroll the whole page instead of the content of the dropdown."
I don't see why we want to be able to switch between scrolling in the whole page and scrolling in the dropdown menu. 🤔

Comment (maltja) [reply src/components/picker/picker.tsx:102]: Yes I agree and will update it 🙂

Comment (adrianschmidt) [reply src/components/picker/picker.tsx:102]: > nice that you found the issues! 🎉 For me both of the versions are broking, I don't know if I do anything wrong when I tested them.

Hmm… I might have linked the wrong versions or something. I'll take a look and post a video showing what I mean.

> "If not all items fit on screen, the user would scroll the whole page instead of the content of the dropdown."
> I don't see why we want to be able to switch between scrolling in the whole page and scrolling in the dropdown menu. 🤔

I think back when that issue was created, we couldn't pop the dropdown _above_ the trigger, so we always just had the space _below_ the trigger to work with. That meant that if the trigger was placed at the bottom of the page, the dropdown would become tiny, sometimes smaller than a single list-item in height. Coupled with the fact that back then, when a dropdown was visible, the page behind it couldn't be scrolled, and we couldn't place the dropdown above the item instead of below it, this led to some workarounds that perhaps doesn't seem very obvious today 😄

Today though, all of those compounding problems have been fixed, and we should never end up in the situation that the dropdown is opened downward and shrunk to less than a single item in height, because we'll just be popping the dropdown upwards instead 👍

Comment (adrianschmidt) [reply src/components/picker/picker.tsx:102]: The first example shown is v30.0.1 where the dropdown becomes longer than the page (thus what the prop is supposed to do), and the second example is v30.0.2 where the dropdown is shorter and gets a scrollbar.

https://user-images.githubusercontent.com/154725/106133121-1f522380-6165-11eb-84b5-dda4173101e2.mov

To clarify, we should still remove the behaviour in the first example 🙂

File: src/components/menu-surface/menu-surface.tsx [hunk 2]
```diff
@@ -176,30 +164,4 @@ export class MenuSurface {
             this.dismiss.emit();
         }
     }
-
-    private ensureMenuFitsInViewPort() {
```
Comment (adrianschmidt) [inline src/components/menu-surface/menu-surface.tsx:180]: > refactor(picker): move calculation of dropdown menu height from menu-surface to picker

Hmm… I haven't looked at the code yet, but I think we'll be needing this calculation in all dropdowns, not just those belonging to limel-picker. So if it can't be done in `menu-surface`, we might want to extract it to a helper or something, that can be used in limel-picker, limel-menu, etc.

Again, I haven't looked at your changes yet, I just saw the commit message and wanted to make a note of it before logging off for the day.

**Comment by maltja** [comment]
I fixed the commit messages that was wrong and confusing 😅. I also added the suggested changes to the deprecated displayFullList property.

File: src/components/picker/picker.tsx [hunk 3]
```diff
@@ -99,6 +97,10 @@ export class Picker {
 
     /**
      * True if the dropdown list should be displayed without cutting the content
+     *
+     * @deprecated This was used for a workaround, and isn't needed any
+     * longer. Setting it has no effect, and the property will be removed
+     * in the next major version.
```
Comment (adrianschmidt) [inline src/components/picker/picker.tsx:103]: We don't need to set this commit as a breaking change, since the functionality was broken in an earlier version, and all we do in this PR is fixing it the bug, albeit in a way that makes the property unnecessary.

The breaking change comes when we actually remove the property.

I can update the commit message, since I have the branch checked out anyway.

Comment (maltja) [reply src/components/picker/picker.tsx:103]: ah okey, thanks! 🙂

File: src/components/picker/picker.tsx [hunk 4]
```diff
@@ -26,6 +26,7 @@ import { createRandomString } from '../../util/random-string';
 const SEARCH_DEBOUNCE = 500;
```
Comment (adrianschmidt) [inline src/components/picker/picker.tsx:26]: It feels like the changes to `menu-surface.tsx` and `portal.tsx` in this commit belong together (and might belong together with the first commit), while the changes to `picker.tsx` feel like they might belong in their own commit.

I assume the changes to `picker.tsx` are needed as a result of moving the dropdown height calculation from menu-surface to portal?

File: src/components/picker/picker.tsx [hunk 5]
```diff
@@ -349,17 +345,20 @@ export class Picker {
...
-                containerStyle={styling}
                 inheritParentWidth={true}
             >
                 <limel-menu-surface
                     open={!!content}
-                    style={{ '--menu-surface-width': '100%' }}
+                    style={{
+                        '--menu-surface-width': '100%',
+                        'max-height': 'inherit',
+                        display: 'flex',
```
Comment (adrianschmidt) [inline src/components/picker/picker.tsx:350]: I was just about to comment that these should be set in the scss file, but that doesn't work because of the portal I guess?

File: src/components/picker/picker.scss [hunk 6]
```diff
@@ -2,4 +2,5 @@
 
 :host {
     position: relative;
+    display: block;
 }
```
Comment (adrianschmidt) [inline src/components/picker/picker.scss:6]: When setting the `display` property of the `:host` selector, that will override the browser styles that hides the element if the attribute `hidden` is set. So we need to add that functionality back:

```suggestion
}

:host([hidden]) {
    display: none;
}
```

File: src/components/menu/menu.scss [hunk 7]
```diff
@@ -28,19 +28,4 @@ $menu-item-background-color: #ebebeb;
```
Comment (adrianschmidt) [inline src/components/menu/menu.scss:28]: The changes to `menu.scss` and `portal.scss` in `fix(portal): give portal exact size of its parent` seem like they might belong together with the changes to `portal.tsx` and `menu-surface.tsx` in the first two commits.

File: src/components/select/select.scss [hunk 8]
```diff
@@ -11,6 +11,12 @@
     display: none;
```
Comment (adrianschmidt) [inline src/components/select/select.scss:11]: I like that all the changes to limel-select are contained in a single commit. I think all the changes to limel-picker also belong in a single commit, and the changes to limel-portal, limel-menu, and limel-menu-surface belong in a single commit (which should come before the others).

I'll move stuff around a little on a separate branch, and you can see if you like it better my way or not 🙂

File: src/components/select/select.scss [hunk 9]
```diff
@@ -11,6 +11,12 @@
...
 }
 
+:host {
+    display: block;
+    position: relative;
+    height: $height-of-mdc-text-field;
+}
```
Comment (adrianschmidt) [inline src/components/select/select.scss:18]:
```suggestion
}

:host([hidden]) {
    display: none;
}
```

File: src/components/picker/picker.tsx [hunk 10]
```diff
@@ -26,6 +26,7 @@ import { createRandomString } from '../../util/random-string';
 const SEARCH_DEBOUNCE = 500;
 const CHIP_SET_TAG_NAME = 'limel-chip-set';
 const ITEM_LIMIT_NO_SCROLL = 5;
+const MAX_HEIGHT_DROP_DOWN = 320;
```
[the remark this answers is not in the document]
Comment (maltja) [reply src/components/picker/picker.tsx]: Yes, I don't know how to fix that 😕 Do you know how I can fix that? 😅

File: src/components/picker/picker.tsx (hunk 4 above)
Comment (maltja) [reply src/components/picker/picker.tsx:26]: yes, I agree 🙂

File: src/components/picker/picker.tsx (hunk 5 above)
Comment (maltja) [reply src/components/picker/picker.tsx:350]: I tried to fix this with Kia but we didn't manage to get it to work. It's because it is set to the container so it is not in the right place in the HTML to be in the Portal's scss file. But I guess we could set it in the portal.tsx file 🤔

File: src/components/picker/picker.scss (hunk 6 above)
Comment (maltja) [reply src/components/picker/picker.scss:6]: ah okey, what does that mean exactly? That it is not possible to hide the element with my changes? 🤔

File: src/components/select/select.scss (hunk 8 above)
Comment (maltja) [reply src/components/select/select.scss:11]: I agree 🙂

File: src/components/picker/picker.tsx (hunk 5 above)
Comment (adrianschmidt) [reply src/components/picker/picker.tsx:350]: No, don't worry about it. It's much better that it's set here than in the portal.

File: src/components/picker/picker.scss (hunk 6 above)
Comment (adrianschmidt) [reply src/components/picker/picker.scss:6]: No, you can still hide it anyway, but if you write

```html
<p hidden>This should be hidden.</p>
```

that paragraph will indeed be hidden. However, if you have styled it with `display:block` via css, and _then_ add the `hidden` attribute, the css wins. Which perhaps wouldn't normally matter much, but since we're building a component library, that's meant to be used in many circumstances by different people, we want our components to behave like native components (within reason).

File: src/components/select/select.scss (hunk 9 above)
Comment (adrianschmidt) [reply src/components/select/select.scss:18]: Ah, this one already had this the `:host([hidden])`, but it comes before this new code. It shouldn't matter for the rendering, but consistency is nice when reading the code, so I'll swap the order around 🙂

File: src/components/picker/picker.scss (hunk 6 above)
Comment (adrianschmidt) [reply src/components/picker/picker.scss:6]: :zap:

File: src/components/picker/picker.tsx (hunk 4 above)
Comment (adrianschmidt) [reply src/components/picker/picker.tsx:26]: :zap:

File: src/components/picker/picker.tsx (hunk 3 above)
Comment (adrianschmidt) [reply src/components/picker/picker.tsx:103]: ⚡

File: src/components/select/select.scss (hunk 9 above)
Comment (adrianschmidt) [reply src/components/select/select.scss:14]: ⚡

File: src/components/menu/menu.scss (hunk 7 above)
Comment (adrianschmidt) [reply src/components/menu/menu.scss:28]: ⚡

File: src/components/picker/picker.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (adrianschmidt) [reply src/components/picker/picker.tsx]: ⚡

File: src/components/select/select.scss (hunk 8 above)
Comment (adrianschmidt) [reply src/components/select/select.scss]: ⚡

**Review by adrianschmidt — APPROVED** [review]
